### ISSUE TITLE
Question regarding self conditioning

### ISSUE THREAD

**Issue by kdwonn** [nicolas-dufour/CAD#2]
Hi, thanks for the code & great work!

I'm struggling to get RIN to work and found your repo really helpful. 

I have two questions regarding self-conditioning:

1. It seems like there was a bug in torch.amp in previous versions, and your implementation is a workaround to follow RIN's self-conditioning while using AMP, without any technical difference. Is my understanding correct?  https://github.com/nicolas-dufour/CAD/blob/bee7ccdfdd83f655141aec693f762af9ff90bc61/cad/models/losses.py#L209
2.  Have you tried training the model without detach() in self-conditioning? I guess it should work fairly well - just wondering if you have any experience with this approach.

Thanks!

**Comment by kdwonn** [comment]
Oh, it seems like amp + no_grad issue is still present (https://github.com/pytorch/pytorch/pull/113685). I guess this is the reason why my RIN implementation hasn't worked so far!

**Comment by nicolas-dufour** [comment]
Hi @kdwonn ,
Indeed, it's the only way to make it work. Otherwise, it deactivates all the mixed precision parameters and you end up training with only the layer norms.
I haven't tried to train without detach, but i wouldn't bet on it working as it would optimize the non-cond latents to help the cond latents. I would bet it breaks the boostraping.

**Comment by kdwonn** [comment]
Thank you so much! I removed the torch.no_grad context and then finally my loss curve started to lower! Regarding detach, I see your point, and detaching gradient seems to be natural.

One more question - it seems like using LAMB optimizer is strongly recommended (https://github.com/lucidrains/recurrent-interface-network-pytorch/issues/2#issuecomment-1407456914). Was this also critical in your implementation?

**Comment by nicolas-dufour** [comment]
@kdwonn it's not critical but it gives a performance boost.
